File: Cipher/examples/test_crypto_opcodes.java
/*
 * Test suite for generated cryptographic operations
 * Validates correctness against known test vectors
 */

import org.junit.Test;
import static org.junit.Assert.*;

public class TestCryptoOpcodes {

    @Test
    public void testRotL32() {
        // Rotate left (circular left shift) for 32-bit values
        // Basic rotation
        assertEquals(0x23456781, CryptoOpcodes.RotL32(0x12345678, 0x04));
        // MSB rotation
        assertEquals(0x01, CryptoOpcodes.RotL32(0x80000000, 0x01));
        // Maximum rotation
        assertEquals(0x80000000, CryptoOpcodes.RotL32(0x01, 0x1F));
    }

    @Test
    public void testPack32BE() {
        // Pack 4 bytes into a 32-bit word (big-endian)
        // Standard packing
        assertEquals(0x12345678, CryptoOpcodes.Pack32BE(0x12, 0x34, 0x56, 0x78));
        // All ones
        assertEquals(0xFFFFFFFF, CryptoOpcodes.Pack32BE(0xFF, 0xFF, 0xFF, 0xFF));
        // Single bit
        assertEquals(0x01, CryptoOpcodes.Pack32BE(0x00, 0x00, 0x00, 0x01));
    }

    @Test
    public void testXorArrays() {
        // XOR two byte arrays
        // Basic XOR
        assertArrayEquals({{0xFF, 0xFF, 0xFF}}, CryptoOpcodes.XorArrays({{0x00, 0xFF, 0xAA}}, {{0xFF, 0x00, 0x55}}));
        // Different lengths
        assertArrayEquals({{0x44, 0x4C}}, CryptoOpcodes.XorArrays({{0x12, 0x34}}, {{0x56, 0x78, 0x9A}}));
    }
}

File: Cipher/examples/test_crypto_opcodes.py
/*
 * Test suite for generated cryptographic operations
 * Validates correctness against known test vectors
 */

import unittest
from crypto_opcodes import *

class TestCryptoOpcodes(unittest.TestCase):

    def test_rotl32(self):
        """Test Rotate left (circular left shift) for 32-bit values"""
        # Basic rotation
        result = RotL32(0x12345678, 0x04)
        self.assertEqual(result, 0x23456781)
        # MSB rotation
        result = RotL32(0x80000000, 0x01)
        self.assertEqual(result, 0x01)
        # Maximum rotation
        result = RotL32(0x01, 0x1F)
        self.assertEqual(result, 0x80000000)

    def test_pack32be(self):
        """Test Pack 4 bytes into a 32-bit word (big-endian)"""
        # Standard packing
        result = Pack32BE(0x12, 0x34, 0x56, 0x78)
        self.assertEqual(result, 0x12345678)
        # All ones
        result = Pack32BE(0xFF, 0xFF, 0xFF, 0xFF)
        self.assertEqual(result, 0xFFFFFFFF)
        # Single bit
        result = Pack32BE(0x00, 0x00, 0x00, 0x01)
        self.assertEqual(result, 0x01)

    def test_xorarrays(self):
        """Test XOR two byte arrays"""
        # Basic XOR
        result = XorArrays([0x00, 0xFF, 0xAA], [0xFF, 0x00, 0x55])
        self.assertEqual(result, [0xFF, 0xFF, 0xFF])
        # Different lengths
        result = XorArrays([0x12, 0x34], [0x56, 0x78, 0x9A])
        self.assertEqual(result, [0x44, 0x4C])

if __name__ == "__main__":
    unittest.main()